prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -515,3 +515,78 @@ void rxr_pkt_handle_eor_recv(struct rxr_ep *ep,
 	rxr_pkt_entry_release_rx(ep, pkt_entry);
 }
 
+
+/* atomrsp packet related functions: init, handle_sent, handle_send_completion and recv
+ *
+ * initialize atomic response packet by creating a packet that hold original data
+ * in rx_entry->iov. rx_entry->iov will then be changed by atomic operation.
+ * release that packet entry until it is sent.
+ */
+int rxr_pkt_init_atomrsp(struct rxr_ep *ep, struct rxr_rx_entry *rx_entry,
+			 struct rxr_pkt_entry *pkt_entry)
+{
+	size_t pkt_size;
+	struct rxr_atomrsp_hdr *atomrsp_hdr;
+
+	assert(rx_entry->atomrsp_pkt);
+	pkt_size = rx_entry->atomrsp_pkt->pkt_size;
+	pkt_entry->pkt_size = pkt_size;
+	pkt_entry->addr = rx_entry->addr;
+	pkt_entry->x_entry = rx_entry;
+
+	atomrsp_hdr = (struct rxr_atomrsp_hdr *)pkt_entry->pkt;
+	atomrsp_hdr->type = RXR_ATOMRSP_PKT;
+	atomrsp_hdr->version = RXR_PROTOCOL_VERSION;
+	atomrsp_hdr->flags = 0;
+	atomrsp_hdr->tx_id = rx_entry->tx_id;
+	atomrsp_hdr->rx_id = rx_entry->rx_id;
+	atomrsp_hdr->seg_size = ofi_total_iov_len(rx_entry->iov, rx_entry->iov_count);
+
+	assert(RXR_ATOMRSP_HDR_SIZE + atomrsp_hdr->seg_size < ep->mtu_size);
+
+	/* rx_entry->atomrsp_buf was filled in rxr_pkt_handle_req_recv() */
+	memcpy((char*)pkt_entry->pkt + RXR_ATOMRSP_HDR_SIZE, rx_entry->atomrsp_buf, atomrsp_hdr->seg_size);
+	pkt_entry->pkt_size = RXR_ATOMRSP_HDR_SIZE + atomrsp_hdr->seg_size;
+	return 0;
+}
+
+void rxr_pkt_handle_atomrsp_sent(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+}
```

Comment:
What's happening here? I saw a comment elsewhere about releasing a packet entry here.

Reply:
I fixed that comment, they are released in rxr_pkt_handle_atomrsp_send_completion()